From 9d48d7bfa5cf534fa6df6f692c22b18afd6c468c Mon Sep 17 00:00:00 2001
From: cxlin <cxlin@zerok.io>
Date: Mon, 28 May 2018 23:07:07 +1000
Subject: [PATCH 2/2] Changed eosiocpp script to dump command db
 (compile_command.json) for toolings.

The following Usage should work:
$ eosiocpp -o hello.wast hello.cpp
$ clang-tidy hello.cpp
---
 tools/eosiocpp.in | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/tools/eosiocpp.in b/tools/eosiocpp.in
index 4dad004..b3e4b8e 100755
--- a/tools/eosiocpp.in
+++ b/tools/eosiocpp.in
@@ -35,21 +35,33 @@ function build_contract {
 
     ($PRINT_CMDS; mkdir $workdir/built)
 
+    ($PRINT_CMDS; echo "[" > $workdir/compile_commands.json)
+
     for file in $@; do
         name=`basename $file`
         filePath=`dirname $file`
+	# use absolute path for the command db
+	if [[ $filePath == "." ]]; then
+	    filePath=`pwd`
+	    file=$filePath/$name
+	fi
 
-        ($PRINT_CMDS; @WASM_CLANG@ -emit-llvm -O3 --std=c++14 --target=wasm32 -nostdinc \
+	command=`echo @WASM_CLANG@ -emit-llvm -O3 --std=c++14 --target=wasm32 -nostdinc \
                                    -nostdlib -nostdlibinc -ffreestanding -nostdlib -fno-threadsafe-statics -fno-rtti \
                                    -fno-exceptions -I ${EOSIO_INSTALL_DIR}/include \
                                    -I${EOSIO_INSTALL_DIR}/include/libc++/upstream/include \
                                    -I${EOSIO_INSTALL_DIR}/include/musl/upstream/include \
                                    -I${BOOST_INCLUDE_DIR} \
                                    -I $filePath \
-                                   -c $file -o $workdir/built/$name
-        )
+                                   -c $file -o $workdir/built/$name`
+
+        ($PRINT_CMDS; $command)
+
+	# record command for toolings
+        (echo "{\"directory\":\"$filePath\",\"command\":\"$command\",\"file\":\"$file\"}," >> $workdir/compile_commands.json)
 
     done
+    (echo "]" >> $workdir/compile_commands.json)
 
     ($PRINT_CMDS; @WASM_LLVM_LINK@ -only-needed -o $workdir/linked.bc $workdir/built/* \
                                    ${EOSIO_INSTALL_DIR}/usr/share/eosio/contractsdk/lib/eosiolib.bc \
@@ -62,6 +74,10 @@ function build_contract {
     ($PRINT_CMDS; ${EOSIO_INSTALL_DIR}/bin/eosio-s2wasm -o $outname -s 16384 $workdir/assembly.s)
     ($PRINT_CMDS; ${EOSIO_INSTALL_DIR}/bin/eosio-wast2wasm $outname ${outname%.*}.wasm -n)
 
+    # move command db to output path
+    outPath=`dirname $outname`
+    ($PRINT_CMDS; mv $workdir/compile_commands.json $outPath/)
+
     ($PRINT_CMDS; rm -rf $workdir)
     set +e
 }
-- 
2.7.4

